web/cgi-bin/DivinumOfficium/Directorium.pm

```diff
@@ -0,0 +1,195 @@
+package DivinumOfficium::Directorium;
+
+use strict;
+use warnings;
+
+use FindBin qw($Bin);
+use lib "$Bin/..";
+
+use DivinumOfficium::FileIO qw(do_read);
+use DivinumOfficium::Date qw(leapyear geteaster);
+
+BEGIN {
+  require Exporter;
+  our $VERSION = 1.00;
+  our @ISA = qw(Exporter);
+  our @EXPORT_OK = qw(get_kalendar get_transfer_table get_stransfer get_tempora_table check_coronatio);
+}
+
+### private vars
+
+my $datafolder = "$Bin/../../www/horas/Latin/Tabulae";
+my %_data;
+my %_dCACHE; # cache everything mainly for kalendar.pl
+
+### private functions
+
+sub load_data_data {
+  my(@lines) = do_read("$datafolder/data.txt");
+  if (@lines == 0) { die "Can't open $datafolder/data.txt"; }
+  shift @lines;
+  foreach (@lines) {
+    my($ver,$kal,$tra,$str) = split(/,/);
+    $_data{$ver}{kalendar} = $kal;
+    $_data{$ver}{transfer} = $tra;
+    $_data{$ver}{stransfer} = $str;
+  }
+  $_dCACHE{loaded} = 1;
+}
+
+sub is_cached {
+  load_data_data unless defined $_dCACHE{loaded};
+  my $key = shift;
+
+  defined $_dCACHE{$key}
+}
+
+sub load_transfer_file {
+  my $name = shift;
+  my $filter = shift;
+  my $type = shift;
```

Comment:
I think it would be worth sanitising these parameters, even though we expect them to come from the datafile.  Maybe enforce that each matches `/[a-z0-9]/i`?

Reply:
In next PR, there will be checks of $version in callers. I suspect it suffice.